### PR TITLE
sever, schedule: record split time 

### PR DESCRIPTION
**Pull request by Connor1996**
merge checker will check the interval between now and last split time to decide whether merge or not.

### CODE REVIEW

File: server/grpc_service.go [hunk 1]
```diff
@@ -342,7 +342,6 @@ func (s *Server) RegionHeartbeat(stream pdpb.PD_RegionHeartbeatServer) error {
 		storeLabel := strconv.FormatUint(storeID, 10)
 
 		regionHeartbeatCounter.WithLabelValues(storeLabel, "report", "recv").Inc()
-		regionHeartbeatLatency.WithLabelValues(storeLabel).Observe(float64(time.Now().UnixNano()/int64(time.Millisecond)) - float64(request.GetTimestamp()))
```
Comment (disksing) [inline server/grpc_service.go]: Why remove it?

File: server/cache.go [hunk 2]
```diff
@@ -409,6 +414,12 @@ func (c *clusterInfo) handleRegionHeartbeat(region *core.RegionInfo) error {
 		}
 		if r.GetVersion() > o.GetVersion() {
 			log.Infof("[region %d] %s, Version changed from {%d} to {%d}", region.GetId(), core.DiffRegionKeyInfo(origin, region), o.GetVersion(), r.GetVersion())
+			// when split or merge, version increases
+			// if key range shrinks, region must has split.
+			if bytes.Compare(region.Region.StartKey, origin.Region.StartKey) > 0 ||
+				(bytes.Compare(region.Region.EndKey, origin.Region.EndKey) < 0 && bytes.Compare(region.Region.EndKey, []byte("")) != 0) {
```
Comment (disksing) [inline server/cache.go]: `bytes.Compare(region.Region.EndKey, []byte("")) != 0` -> `len(key) > 0`

File: server/config.go [hunk 3]
```diff
@@ -409,11 +412,28 @@ func (c *ScheduleConfig) clone() *ScheduleConfig {
 	}
 }
 
+const (
+	defaultMaxReplicas          = 3
+	defaultMaxSnapshotCount     = 3
+	defaultMaxPendingPeerCount  = 16
+	defaultMaxMergeRegionSize   = 0
+	defaultSplitMergeInterval   = 1 * time.Minute
```
[the remark this answers is not in the document]
Comment (disksing) [reply server/config.go]: How about 30min?

File: server/schedule/merge_checker.go [hunk 4]
```diff
@@ -52,6 +54,12 @@ func (m *MergeChecker) Check(region *core.RegionInfo) (*Operator, *Operator) {
 		return nil, nil
 	}
 
+	last := time.Unix(region.GetLastSplitTimestamp(), 0)
+	if time.Since(last) < m.cluster.GetSplitMergeInterval() {
+		checkerCounter.WithLabelValues("merge_checker", "too_frequent").Inc()
```
Comment (disksing) [inline server/schedule/merge_checker.go]: `too_frequent` -> `recently-split`?

Comment (disksing) [reply server/schedule/merge_checker.go]: recently_split

File: server/cache.go [hunk 5]
```diff
@@ -409,6 +414,12 @@ func (c *clusterInfo) handleRegionHeartbeat(region *core.RegionInfo) error {
 		}
 		if r.GetVersion() > o.GetVersion() {
 			log.Infof("[region %d] %s, Version changed from {%d} to {%d}", region.GetId(), core.DiffRegionKeyInfo(origin, region), o.GetVersion(), r.GetVersion())
+			// when split or merge, version increases
+			// if key range shrinks, region must has split.
```
Comment (disksing) [inline server/cache.go]: must have.

**Comment by disksing** [comment]
@huachaohuang You may want to increase the `split-merge-interval` config in lightning before importing data.

File: vendor/github.com/pingcap/kvproto/pkg/metapb/metapb.pb.go [hunk 6]
```diff
@@ -181,10 +181,11 @@ func (m *RegionEpoch) GetVersion() uint64 {
...
 	// Region key range [start_key, end_key).
-	StartKey    []byte       `protobuf:"bytes,2,opt,name=start_key,json=startKey,proto3" json:"start_key,omitempty"`
-	EndKey      []byte       `protobuf:"bytes,3,opt,name=end_key,json=endKey,proto3" json:"end_key,omitempty"`
-	RegionEpoch *RegionEpoch `protobuf:"bytes,4,opt,name=region_epoch,json=regionEpoch" json:"region_epoch,omitempty"`
-	Peers       []*Peer      `protobuf:"bytes,5,rep,name=peers" json:"peers,omitempty"`
+	StartKey           []byte       `protobuf:"bytes,2,opt,name=start_key,json=startKey,proto3" json:"start_key,omitempty"`
+	EndKey             []byte       `protobuf:"bytes,3,opt,name=end_key,json=endKey,proto3" json:"end_key,omitempty"`
+	RegionEpoch        *RegionEpoch `protobuf:"bytes,4,opt,name=region_epoch,json=regionEpoch" json:"region_epoch,omitempty"`
+	Peers              []*Peer      `protobuf:"bytes,5,rep,name=peers" json:"peers,omitempty"`
+	LastSplitTimestamp int64        `protobuf:"varint,6,opt,name=last_split_timestamp,json=lastSplitTimestamp,proto3" json:"last_split_timestamp,omitempty"`
```
Comment (nolouch) [inline vendor/github.com/pingcap/kvproto/pkg/metapb/metapb.pb.go]: why did `gopkg.toml` not change?

File: server/config.go [hunk 7]
```diff
@@ -415,11 +418,28 @@ func (c *ScheduleConfig) clone() *ScheduleConfig {
 	}
 }
 
+const (
+	defaultMaxReplicas          = 3
+	defaultMaxSnapshotCount     = 3
+	defaultMaxPendingPeerCount  = 16
+	defaultMaxMergeRegionSize   = 0
+	defaultSplitMergeInterval   = 30 * time.Minute
```
Comment (huachaohuang) [inline server/config.go]: Can we default to 1 hour?
I think 1 hour should be enough for TiDB Lighting in most cases.

File: vendor/github.com/pingcap/kvproto/pkg/metapb/metapb.pb.go (hunk 6 above)
Comment (Connor1996) [reply vendor/github.com/pingcap/kvproto/pkg/metapb/metapb.pb.go]: just copy. When related kvproto PR merged, I will update it.

File: server/config.go (hunk 7 above)
Comment (Connor1996) [reply server/config.go]: okay

**Comment by siddontang** [comment]
PTAL @huachaohuang

**Pull request closed without merging**